OSDOCS-6528: Created documentation for creating a shared VPC (#63793)
openshift/openshift-docs: modules/rosa-sharing-vpc-cluster-creation.adoc

// Module included in the following assemblies:
//
// * networking/rosa-shared-vpc-config.adoc
:_content-type: PROCEDURE
[id="rosa-sharing-vpc-cluster-creation_{context}"]
= Creating your cluster in a shared VPC

> After the cluster-creating AWS account user has created the DNS domain and the IAM roles, you must create a hosted space and update the trust policy on the IAM role that you created for sharing the VPC.

Author: It needs revising, but the issue with this revision is that they haven't created the cluster yet; just the VPC was created.

> After the AWS account user who initially created the DNS domain and the IAM roles, the VPC-owning AWS account user must create a hosted space and update the trust policy on the IAM role that was created for sharing the VPC.

Reviewer: ```suggestion
After the AWS account user who initially created the DNS domain and the IAM roles, the VPC-owning AWS account user must create a private hosted zone and update the trust policy on the IAM role that was created for sharing the VPC.
```

.Procedure

> . The AWS account user who owns the VPC must update the VPC sharing IAM role and add the Installer and Ingress Operator roles to the principal section of the trust policy.

Reviewer 2: @EricPonvelle As this is very important configure steps for Shared-VPC install, can we make it more clear in the doc?

@oriAdler WDYT?

Author: @yunjiang29 @oriAdler Can you elaborate on what we need to make clear? These steps happen in the AWS console, I believe, so there's not a lot I can document from the Red Hat side. 

I could show an updated Trust Policy if that would help.

Reviewer: I think an example can be great, it's a prerequisite for installation, otherwise, the backend preflight will fail, and the cluster creator will have to communicate back to the VPC owner the roles' ARNs.

> . After updating the trust policy, the VPC-owning AWS account user creates a private hosted zone in the link:https://us-east-1.console.aws.amazon.com/route53/v2/[Route 53 section of the AWS console]. The hosted spaces name is `<cluster-name>.<dns_domain>`. The private hosted zone needs to include the created VPC's AWS region and ID.

Reviewer 2: ```suggestion
. After updating the trust policy, the VPC-owning AWS account user creates a private hosted zone in the link:https://us-east-1.console.aws.amazon.com/route53/v2/[Route 53 section of the AWS console]. The hosted spaces name is `<cluster-name>.<dns_domain>`. The private hosted zone needs to be associated with created VPC.
```

> . After the hosted zone is created, provide the following to the AWS account user who creats the cluster:

Reviewer 3: ...account user who 'created' the cluster.

// Module included in the following assemblies:
//
// * networking/rosa-shared-vpc-config.adoc
:_content-type: PROCEDURE
[id="rosa-sharing-vpc-cluster-creation_{context}"]
= Creating your cluster in a shared VPC

The *cluster-creating AWS account* user runs the following command in their terminal to create the cluster in the shared VPC.

[source,terminal]
----
rosa create cluster --cluster-name <cluster_name> --sts --operator-roles-prefix <prefix> --oidc-config-id <oidc_config_id> --region us-east-1 --subnet-ids <subnet_ids> --private-hosted-zone-id <hosted_zone_ID> --shared-vpc-role-arn <vpc-role-arn> --base-domain <dns-domain>
----

[IMPORTANT]
====

> After deleting a shared-VPC cluster, three DNS records would not be removed from the private hosted zone.

Reviewer 4: Should this read "will" instead of "would"? The past tense is confusing here.

Author: @wgordon17 - if I am understanding this issue correctly, these records would sometimes stick so we need to make this note more "this situation is possible."

OR is it a situation where the records never fully delete regardless what you do?

> In situations where the DNS records remain, any subsequent install attempts using the same values would result in installation errors, blocking installs from proceeding and succeeding.

Reviewer 4: Again, replace "would" with "will".

Author: See previous.

// Module included in the following assemblies:
//
// * networking/rosa-shared-vpc-config.adoc
:_content-type: PROCEDURE
[id="rosa-sharing-vpc-cluster-creation_{context}"]
= Creating your cluster in a shared VPC

> The *cluster-creating AWS account* user runs the following command in their terminal to create the cluster in the shared VPC.

Reviewer 5: ```suggestion
To create a cluster in a shared VPC, complete the following steps. 

[NOTE]
====
Installing a cluster in a shared VPC is supported only for OpenShift 4.13.9 and later.
====

.Prerequisites

You must have the *cluster-creating AWS account* user role.

.Procedure

In a terminal, enter the following command to create a cluster in the shared VPC:

```
- This revision might be overkill, but the current format of placing the command between two admonitions (Note and Important) made it a little hard for me to figure out where to look at first. This revision adds a few more visual clues that will hopefully guide the reader well, and places the info that you'd probably want to know before you get started before the task.
- In the note text, I also moved "only" to be as close to the phrase that it modifies as possible, per IBM Style.
- One thing to check: I wasn't sure if bold highlighting was correct for the "cluster-creating AWS account" user name. I've seen user names highlighted in monospace before, but not in bold. Is that typical for ROSA docs?

> [NOTE]

Reviewer 5: If you decide to adopt the revision that I suggested, you can remove this note, since it's been moved further up in the topic.

// Module included in the following assemblies:
//
// * networking/rosa-shared-vpc-config.adoc
:_content-type: PROCEDURE
[id="rosa-sharing-vpc-cluster-creation_{context}"]
= Creating your cluster in a shared VPC

> After the AWS account user who initially created the DNS domain and the IAM roles, the VPC-owning AWS account user must create a private hosted zone and update the trust policy on the IAM role that was created for sharing the VPC.

Reviewer 6: ```suggestion
After the *cluster-creator AWS account* user provides the DNS domain and the IAM roles, the *VPC-owning AWS account* user must create a private hosted zone and update the trust policy on the IAM role that was created for sharing the VPC.
```

.Procedure

> . The AWS account user who owns the VPC must update the VPC sharing IAM role and add the Installer and Ingress Operator roles to the principal section of the trust policy.

Reviewer 6: ```suggestion
. The *VPC-owner AWS account* user must update the VPC sharing IAM role and add the Installer and Ingress Operator roles to the principal section of the trust policy.
```

+
[source,terminal]
----
{
  "Version": "2012-10-17",
  "Statement": [
    {
	  "Sid": "Statement1",
	  "Effect": "Allow",
	  "Principal": {
	  	"AWS": [

>           "arn:aws:iam::<Cluster-Creator's-AWS-Account-ID>:role/ManagedOpenShift-ingress-operator-cloud-credentials",

Reviewer 6: ```suggestion
          "arn:aws:iam::<Cluster-Creator's-AWS-Account-ID>:role/<prefix>-ingress-operator-cloud-credentials",
```

          "arn:aws:iam::<Cluster-Creator's-AWS-Account-ID>:role/ManagedOpenShift-Installer-Role"
        ]            
	  },
	  "Action": "sts:AssumeRole"
	}
  ]
}
----

> . After updating the trust policy, the VPC-owning AWS account user creates a private hosted zone in the link:https://us-east-1.console.aws.amazon.com/route53/v2/[Route 53 section of the AWS console]. The hosted spaces name is `<cluster-name>.<dns_domain>`. The private hosted zone needs to be associated with the created VPC.

Reviewer 6: ```suggestion
. After updating the trust policy, the *VPC-owning AWS account* user creates a private hosted zone in the link:https://us-east-1.console.aws.amazon.com/route53/v2/[Route 53 section of the AWS console]. In the hosted zone configuration, the domain name will be `<cluster-name>.<dns_domain>`. 
. After private hosted zone creation, it will need to be associated with the intended VPC.
```

> . After the hosted zone is created, provide the following to the AWS account user who created the cluster:

Reviewer 6: ```suggestion
. Once the hosted zone is created and associated to the VPC, provide the following to the *cluster-creator AWS account* user:
    * Hosted Zone ID
    * Region
    * Intended subnet IDs
```

+
[source,terminal]
----

> rosa create cluster --cluster-name <cluster_name> --sts --operator-roles-prefix <prefix> --oidc-config-id <oidc_config_id> --region us-east-1 --subnet-ids <subnet_ids> --private-hosted-zone-id <hosted_zone_ID> --shared-vpc-role-arn <vpc-role-arn> --base-domain <dns-domain>

Reviewer 6: I feel like this should be broken into _another_ module, and make it obvious that this is what the *cluster-creator AWS account* user should be running in _their_ terminal.
